src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1120,15 +1132,35 @@ private void OnFileChanged(object sender, FileSystemEventArgs e)
             // - the host.json file was changed
             // - a function.json file was changed
             // - a function directory was added/removed/renamed
+            // A full host shutdown is performed when an assembly (.dll, .exe) in a watched directory is modified
             string fileName = Path.GetFileName(e.Name);
             if (isWatchedDirectory ||
                 ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 !_directorySnapshot.SequenceEqual(Directory.EnumerateDirectories(ScriptConfig.RootScriptPath)))
             {
-                // a host level configuration change has been made which requires a
-                // host restart
-                _restart(e);
+                bool shutdown = false;
+                string fileExtension = Path.GetExtension(fileName);
+                if (string.Equals(fileExtension, ".dll", StringComparison.OrdinalIgnoreCase) || string.Equals(fileExtension, ".exe", StringComparison.OrdinalIgnoreCase))
```

Comment:
Can you use/share AssemblyFileTypes here as well (you used in other places above)